File: rd-kt/rd-framework/src/test/kotlin/com/jetbrains/rd/framework/test/cases/RdIdLegacyCollisionTest.kt
package com.jetbrains.rd.framework.test.cases

import com.jetbrains.rd.framework.*
import com.jetbrains.rd.framework.base.AllowBindingCookie
import com.jetbrains.rd.framework.base.RdBindableBase
import com.jetbrains.rd.framework.test.util.TestScheduler
import com.jetbrains.rd.framework.test.util.TestWire
import com.jetbrains.rd.util.lifetime.LifetimeDefinition
import kotlin.test.Test
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue

class RdIdLegacyCollisionTest {

    private class Leaf : RdBindableBase()

    private class CollisionModel(childCount: Int) : RdBindableBase() {
        val kids: List<Leaf> = (0 until childCount).map { Leaf() }
        init {
            kids.forEachIndexed { i, leaf -> bindableChildren.add("x$i" to leaf) }
        }
    }

    @Suppress("DEPRECATION")
    private fun legacy(): IIdentities = Identities(IdKind.Server)
    private fun sequential(): IIdentities = SequentialIdentities(IdKind.Server)

    @Test
    fun `legacy named child rdids do not collide across parents`() = childRdIdsDoNotCollide(legacy())

    @Test
    fun `SequentialIdentities named child rdids do not collide across parents`() = childRdIdsDoNotCollide(sequential())

    @Suppress("DEPRECATION")
    private fun childRdIdsDoNotCollide(identities: IIdentities) {
        val ld = LifetimeDefinition()
        val lifetime = ld.lifetime

        val protocol = Protocol("Collision", Serializers(), identities, TestScheduler, TestWire(TestScheduler), lifetime)

        val model1 = CollisionModel(31)
        val model2 = CollisionModel(1)

        AllowBindingCookie.allowBind {
            model1.identify(identities, RdId(3L), false)
            model2.identify(identities, RdId(1L), false)

            model1.preBind(lifetime, protocol, "m1")
            model1.bind()
            // Without the fix this is where model2.x0 duplicates model1.x0 and register() throws.
            model2.preBind(lifetime, protocol, "m2")
            model2.bind()
        }

        assertNotEquals(model1.kids[0].rdid.hash, model2.kids[0].rdid.hash,
            "named children of different parents must not collide")

        val all = HashSet<Long>()
        (model1.kids + model2.kids).forEach { assertTrue(all.add(it.rdid.hash), "duplicate child id ${it.rdid.hash}") }

        ld.terminate()
    }
}

File: rd-net/Test.RdFramework/RdIdLegacyCollisionTest.cs
using System;
using System.Collections.Generic;
using JetBrains.Collections.Viewable;
using JetBrains.Lifetimes;
using JetBrains.Rd;
using JetBrains.Rd.Base;
using JetBrains.Rd.Impl;
using JetBrains.Serialization;
using NUnit.Framework;

namespace Test.RdFramework
{
  [TestFixture]
  public class RdIdLegacyCollisionTest
  {
    private sealed class Leaf : RdBindableBase
    {
    }

    private sealed class CollisionModel : RdBindableBase
    {
      public readonly List<Leaf> Kids = new List<Leaf>();

      public CollisionModel(int childCount)
      {
        for (var k = 0; k < childCount; k++)
        {
          var leaf = new Leaf();
          Kids.Add(leaf);
          BindableChildren.Add(new KeyValuePair<string, object>("x" + k, leaf));
        }
      }
    }

    private sealed class StubWire : IWire
    {
      public bool IsStub => true;
      public void Send<TParam>(RdId id, TParam param, Action<TParam, UnsafeWriter> writer) { }
      public void Advise(Lifetime lifetime, IRdWireable entity) { }
      public ProtocolContexts Contexts { get; set; } = null!;
      public IRdWireable? TryGetById(RdId rdId) => null;
    }

    private static IIdentities Legacy()
    {
#pragma warning disable CS0618 // Type or member is obsolete
      return new Identities(IdKind.Server);
#pragma warning restore CS0618
    }

    private static IIdentities Sequential() => new SequentialIdentities(IdKind.Server);

    private static IEnumerable<TestCaseData> Cases()
    {
      yield return new TestCaseData((Func<IIdentities>)Legacy).SetName("legacy Identities (with the fix)");
      yield return new TestCaseData((Func<IIdentities>)Sequential).SetName("SequentialIdentities");
    }

    [TestCaseSource(nameof(Cases))]
    public void ChildRdIdsDoNotCollideAcrossParents(Func<IIdentities> identitiesFactory)
    {
      var ld = new LifetimeDefinition();
      var lifetime = ld.Lifetime;
      var protocol = new Protocol("Collision", new Serializers(), identitiesFactory(),
        SynchronousScheduler.Instance, new StubWire(), lifetime);

      var model1 = new CollisionModel(31);
      var model2 = new CollisionModel(1);

      using (AllowBindCookie.Create())
      {
        model1.Identify(protocol.Identities, new RdId(3), stable: false);
        model2.Identify(protocol.Identities, new RdId(1), stable: false);

        model1.PreBind(lifetime, protocol, "m1");
        model1.Bind();
        // Without the fix this is where model2.x0 duplicates model1.x0 and Register throws.
        model2.PreBind(lifetime, protocol, "m2");
        model2.Bind();
      }

      Assert.AreNotEqual(model1.Kids[0].RdId.Value, model2.Kids[0].RdId.Value,
        "named children of different parents must not collide");

      var all = new HashSet<long>();
      foreach (var leaf in model1.Kids) Assert.True(all.Add(leaf.RdId.Value), "duplicate id in model1");
      foreach (var leaf in model2.Kids) Assert.True(all.Add(leaf.RdId.Value), "duplicate id in model2");

      ld.Terminate();
    }
  }
}
